text/0103-introduce-core-index-commitment.md

```diff
@@ -0,0 +1,236 @@
+	/// Thus it cannot be derived entirely from the relay-parent.
+	persisted_validation_data_hash: Hash,
+	/// The blake2-256 hash of the PoV.
+	pov_hash: Hash,
+	/// The root of a block's erasure encoding Merkle tree.
+	erasure_root: Hash,
+	/// Reserved bytes.
+	reserved2: [u8; 64],
+	/// Hash of the para header that is being generated by this candidate.
+	para_head: Hash,
+	/// The blake2-256 hash of the validation code bytes.
+	validation_code_hash: ValidationCodeHash,
+}
+```
+
+In future format versions, parts of the `reserved1` and `reserved2` bytes can be used to include additional information in the descriptor.
+
+
+#### Versioned `CandidateReceipt` and `CommittedCandidateReceipt` primitives:
+
+We want to decouple the actual representation of the `CandidateReceipt` from the higher level code. This should make it easier to implement future format versions of this primitive. To hide the logic of versioning the descriptor fields will be private and accessor methods are provided.
+
+``` 
+pub enum VersionedCandidateReceipt {
+	V1(CandidateReceipt),
+	V2(CandidateReceiptV2),
+}
+
+impl VersionedCandidateReceipt {
+	/// Returns the core index the candidate has commited to. 
+	/// Returns `None` if the candidate receipt is the old version (v1).
+	fn core_index() -> Option<CoreIndex>;
+
+	/// Returns the session index of the candidate relay parent.
+	fn session_index() -> Option<CoreIndex>;
+
+	/// ...
+}
+
+```
+
+A manual decode `Decode`  implementation is required to account for version detection and constructing the appropriate variant.
+
+
+#### New primitive for representing the `CoreIndex` commitments.**
+
+```
+pub enum UMPSignal {
+	OnCore(CoreIndex),
+}
+```
+
+### Cumulus primitives
+
+Add a new version of the `ParachainInherentData` structure which includes an additional `core_index` field.
+```
+pub struct ParachainInherentData {
+	pub validation_data: PersistedValidationData,
+	/// A storage proof of a predefined set of keys from the relay-chain.
+	///
+	/// Specifically this witness contains the data for:
+	///
+	/// - the current slot number at the given relay parent
+	/// - active host configuration as per the relay parent,
+	/// - the relay dispatch queue sizes
+	/// - the list of egress HRMP channels (in the list of recipients form)
+	/// - the metadata for the egress HRMP channels
+	pub relay_chain_state: sp_trie::StorageProof,
+	/// Downward messages in the order they were sent.
+	pub downward_messages: Vec<InboundDownwardMessage>,
+	/// HRMP messages grouped by channels. The messages in the inner vec must be in order they
+	/// were sent. In combination with the rule of no more than one message in a channel per block,
+	/// this means `sent_at` is **strictly** greater than the previous one (if any).
+	pub horizontal_messages: BTreeMap<ParaId, Vec<InboundHrmpMessage>>,
+	/// The core index on which the parachain block must be backed
+	pub core_index: CoreIndex,
```

Comment:
I don't think this fully resolves the issue. With this any valid collator could still choose the core index freely. Much smaller issue than before, but still. Also on top of that how do we upgrade the ParachainInherentData?

Ideally we would have a compact proof in above `relay_chain_state` telling us which cores we have available at that relay parent and then the parachain logic just iterates through them. First block goes on core 1, child goes on core 2, ....

Reply:
In any case, I believe we should think about this part (how does the runtime know its core index for the current block?) a bit more.

Reply:
Ok. Just had a late night idea, doing away with both inherent and proofs. We change the UMPSignal name to `CoreSelector` or something similar. What the parachain then does is simply putting the block number of the parachain block in there. Now let's assume for a moment that we know what position in the claim queue the collation is supposed to fill. What we can do then is find all cores the parachain is scheduled and sort them by core number, e.g. 3 cores:

```
[7, 20, 48]
```
Now we take `CoreSelector` % 3 to determine the desired core.

This has a couple of nice properties:

1. No additional proofs required.
2. No additional inherent data required.
3. It autoscales. No matter how many cores there are, we have a proper assignment for each core.

The open question to resolve: How does the parachain tell us (and how would the runtime even know), what position in the claim queue it intents to occupy. Simple and working solution: Don't only pass `CoreSelector`, but also `ClaimQueuePos`.  Then the validators know. Remaining: How should the parachain runtime know? In the most simple case: It is just static, e.g. always position 1. That works and is probably good enough for almost every parachain, but if it is constant, why even pass it? Also most/all parachains will likely pick 1.... 

Reasons to pick other values: 0 ... synchronous backing, if you don't fill the block much and want low latency. 2 ... actually I don't know, maybe because you have slow networking and need the additional time to make sure the collation will make it.

TL;DR: adding `ClaimQueuePos` to the message works. How the parachain would determine anything else than a static value is an open question, but having it in the commitments leaves maximum flexibility.

Reply:
Yeah, good point. Your idea would work, but feels a bit complicated. 

A couple of things I would adjust to make it more simple:
-  Pass the CoreIndex as commitment. It is also a much cleaner interface.
- Use state proofs for the claimqueue for simplicity. They should be made compact.
- The parachain runtime stores a `BTreeSet` of what cores have been used so far at current relay parent. If relay parent advances we clear the used indices.

Reply:
I don't think this is simpler. Also the claim queue, if we consider 100 to 200 cores will have a significant size. Putting that in into each and every block seems to be a non-starter and can easily be avoided by the above. In the end what the para runtime would do is very similar to the algorithm I described for backers, but instead of doing it cost efficient on the backers we would need to do it in the runtime.

Reply:
I have spent some time to think about this and I don't have any better idea. I'll adjust the RFC and use `CoreSelector` and `ClaimQueuePos` commitments.